README
A quick description of the content of this repo.

FILES:

APOD.pm:
	A Perl module for accessing the Astronomy Picture Of the Day as
	well as its description.

apod_get:
	A Perl script which, by using APOD.pm, saves the image and 
	description file in the current directory.

apod_desktop:
	A bash script for setting an annotated version of the APOD image
	as the desktop background. In essence it runs apod_get in a 
	specific directory(~/.APOD/new/) and then annotates the APOD
	description to the image (into ~/.APOD/) via linnea_annotate.
	It then sets the background via any method set in
	background_set() (see the script itself to change it)

linnea_annotate:
	Although this turned out as a part of the APOD code, it's now
	completely unrelated and should probably be moved to its own
	repository.

	It's a basic text annotator. It takes an input image and text,
	annotates the text onto it and produces an output image.
	It can do some helping stuff, like automatically create an output
	image which is the same size as the current desktop.

NOTES:

Most of this stuff assumes that the files are globally installed, meaning
that APOD.pm needs to go into your perl module path.

For apod_desktop to run, apod_desktop and linnea_annotate need to be
globally installed (in, say, /usr/local/bin/ or the like).


font/ChangeLog
2011-06-23 Andrey V. Panov
	Manually instructed several glyphs.
	Changed Win and HHead OS/2 metrics.
	Version 0.4.2 is released.

2011-06-07 Andrey V. Panov
	Added several Latin and Cyrillic extended glyphs, "yen" sign.
	Added OFL version of the fonts.
	Version 0.4.1 is released.

2011-01-18 Andrey V. Panov
	Added proportional and tabular oldstyle, superior figures to all the 
	fonts.
	Added more glyphs, mostly Latin and Cyrillic extended ones.
	Added more vulgar fractions.
	Added kerning in BoldItalic.
	Version 0.4 is released.

2010-10-26 Andrey V. Panov
	All the fonts have manual instructions.
	Added proportional and tabular oldstyle figures to Regular.
	Added more glyphs, mostly Latin and Cyrillic extended ones.
	Changed spacing of small capitals.
	Version 0.3 is released.

2010-04-29 Andrey V. Panov

	Added Bold Italic font.
	Regular font is mostly manually instructed.
	The second beta version (0.2) is released.

2009-08-14 Andrey V. Panov
	
	Tuned several kerning pairs.

2009-08-09 Andrey V. Panov

	Added "copyright" sign (U+A9), "cedilla", "Ccedilla", "ccedilla".
	The first beta version (0.1) is released.

2009-08-07 Andrey V. Panov

	Added old style kerning ('kern' table).

2009-07-31 Andrey V. Panov

	Added kerning to basic Latin and Cyrillic in Italic.
	Added kerning to small caps in Regular.

2009-07-30 Andrey V. Panov

	Added kerning to basic Latin and Cyrillic in Bold.
	Changed design of several lowercase letters to more curly style in 
	Italic. The previous variants are saved as stylistic set 01.

2009-07-27 Andrey V. Panov

	Added "fi" and "fl" ligatures.

2009-07-18 Andrey V. Panov

	Added kerning to basic Latin and Cyrillic in Regular.

2009-07-17 Andrey V. Panov

	Added small caps to Regular.

2009-07-01 Andrey V. Panov

	Added substitutions for tabular numbers and signs.

2009-06-29 Andrey V. Panov

	Added Bold font.

2009-06-23 Andrey V. Panov

	Manually instructed U+414, U+41B, U+434, U+43B, U+2116 in Regular.

2009-06-22 Andrey V. Panov

	Manually instructed "W", "w" in Regular.

2009-06-10 Andrey V. Panov

	Manually instructed glyphs with diagonals excepting "W", "w" in Regular.

2009-01-09 Andrey V. Panov

	Added Italic font.

2008-12-21 Andrey V. Panov

	Enlarged x-height by 10 em units.

2008-12-02 Andrey V. Panov

	Finished more or less basic set of glyphs in Regular font.

test.pl
#/usr/bin/env perl
# Linnea.
#
use warnings;
use strict;

use Carp;
use Getopt::Long;

use X11::Resolution;

use Imager;
use Imager::Fill;
use Imager::Font::Wrap;
use Data::Dumper;

my %opts = (
	"text_bg"          => '0a0a10',
	"text_fg"          => 'ffffff',
	"text"             => '',
	"text_file"        => '',
	"font_pts"         => 14,
	"font"             => './font/Istok-Regular.ttf',
	"screen"           => 0,
	"resolution"       => '0x0',
	"text_bg_blur"     => 4.0,
	"text_bg_contrast" => 9.0,
	"bg"               => 9,
	"valign"           => 'top',    # top center bottom
	"halign"           => 'center', # left center right
	"justify"          => 'fill',   # left center right fill
);

sub help {

	printf("Usage: %s [args] <input-file> <output-file>
  --text-bg           text backdrop coloring (default '%s')
  --text-fg           text forgeground color (default '%s')
  --text              text                   (default '%s')
  --text-file         file containing text   (default '%s')
  --font-pts          point size of font     (default '%s')
  --font              font file              (default '%s')
  --screen            screen number          (default %d)
  --resolution        resolution   (default use desktop resolution)
  --text-bg-blur      bg gaussian blur intensity (default %f)
  --text-bg-contrast  bg contrast intensity      (default %f)
  --bg                bg solid color             (default none)
  --valign            vertical text alignment    (default '%s')
  --halign            horizontal text alignment  (default '%s')
  --justify           text justification method  (default '%s')
", 
	$0,
	$opts{text_bg},
	$opts{text_fg},
	$opts{text},
	$opts{text_file},
	$opts{font_pts},
	$opts{font},
	$opts{screen},
	$opts{text_bg_blur},
	$opts{text_bg_contrast},
	$opts{valign},
	$opts{halign},
	$opts{justify});

}

# Input : 6-character hex string (0a1fff) 
# Return: decimal rgb-triplet array 
sub color_parse {
	my $color = lc shift;
	if(length($color) != 6) {
		croak "Bad color '$color': " . 
		      "format is rrggbb where each 'xx'-pair is a hex-value";
	}
	my @colors;
	push @colors, hex substr($color, $_, 2) for (0, 2, 4);

	return @colors;
}

sub resolution_parse {
	my $in = lc shift;
	my @out = split /x/, $in;
	for (@out) {
		croak "Bad resolution '$in': format is WxH" if /[^\d]/;
	}
	return @out;
}

sub resolution_get {
	my $screen = shift;

	my $X = new X11::Resolution;

	if($X->{error}){
		croak "Can't instantiate X11::Resolution: " . $X->{errorString};
	}

	my ($w, $h) = $X->getResolution($screen);

	if($X->{error}){
		croak "Can't get X resolution: " . $X->{errorString};
	}

	return($w, $h);
}
##############################################################################

my $result = GetOptions ("text-bg=s"          => \$opts{text_bg},
	                 "text-fg=s"          => \$opts{text_fg},
	                 "text=s"             => \$opts{text},
	                 "text-file=s"        => \$opts{text_file},
	                 "font-pts=i"         => \$opts{text_pts},
	                 "font=s"             => \$opts{font},
	                 "screen=i"           => \$opts{screen},
	                 "resolution=s"       => \$opts{resolution},
	                 "text-bg-blur=f"     => \$opts{text_bg_blur},
	                 "text-bg-contrast=f" => \$opts{text_bg_contrast},
	                 "bg=s"               => \$opts{bg},
	                 "valign=s"           => \$opts{valign},
	                 "valign=s"           => \$opts{valign},
	                 "justify=s"          => \$opts{valign},
			 );

if(@ARGV != 2) {
	help();
	exit;
}

# Container hashes
my %src = (
	w    => 0,
	h    => 0,
	data => undef,
);
my %txt = (
	w    => 0,
	h    => 0,
	data => undef,
	color=> undef,
	font => undef,
);
my %dst = (
	file => '',
	w    => 0,
	h    => 0,
	data => undef,
);

($src{file}, $dst{file}) = @ARGV;
my @text_color_fg        = color_parse($opts{text_fg});
my @text_color_bg        = color_parse($opts{text_bg});
($dst{w}, $dst{h})       = resolution_parse($opts{resolution});

if($opts{text_file} ne '') {
	
	if($opts{text} ne '') {
		print "Overriding supplied --text with data in " .
		      "\"$opts{text_file}\"\n";
	}

	$opts{text} = '';

	open F, '<', $opts{text_file} 
		or croak "Can't open \"$opts{text_file}\": $!";

	$opts{text} .= $_ for <F>; 

	close F;

}

my $use_desktop_size = 0;

# If either width or height is zero, use desktop size.
if($dst{w} == 0 || $dst{h} == 0) {
	$use_desktop_size = 1;
	($dst{w}, $dst{h}) = resolution_get($opts{screen});
}

printf("In  : \"%s\"\n".
       "Out : \"%s\"\n". 
       "Size: %dx%d %s\n", 
       $src{file}, 
       $dst{file},
       $dst{w}, 
       $dst{h},
       $use_desktop_size ? '(desktop size)' : ''
       );


##############################################################################

# (old components - still needs work!)
# TODO:
# - clean up variable names
# - clean up error messages
# - adjust to use above opts
# .. etc

# Load source image
$src{data} = Imager->new(file     => $src{file},
			 channels => 4) or die Imager->errstr;

($src{w}, $src{h}) = ($src{data}->getwidth(), 
		      $src{data}->getheight());

# Create empty destination canvas
$dst{data} = Imager->new(xsize    => $dst{w},
			 ysize    => $dst{h},
			 channels => 4) or die Imager->errstr;

# Scale it to fit height if necessary
if($src{h} > $dst{h}) {
	my $tmp = $src{data}->scale(ypixels => $dst{h}) or die Imager->errstr;
	$src{data} = $tmp;
}

($src{w}, $src{h}) = ($src{data}->getwidth(), 
		      $src{data}->getheight());

# Paste scaled image to center of canvas
# TODO take $opts{halign} into account
my $offs_x = (.5 * $dst{w}) - (.5 * $src{w});
my $offs_y = (.5 * $dst{h}) - (.5 * $src{h});

$dst{data}->paste(left => $offs_x,
		  top  => $offs_y,
		  src  => $src{data}) or die $dst{data}->errstr;

$txt{color} = Imager::Color->new(@text_color_bg);
$txt{font} = Imager::Font->new(file  => $opts{font},
			       size  => $opts{font_pts},
			       color => $txt{color}) or die Imager->errstr;

# Calculate size of annotation
# Note that we cannot know the actual width of the text;
# we can only pass our desired maximum width and this is what we get.

# Padding for blurring to dissipate without clipping.
# TODO cleanup
my $pad_x = 30;
my $pad_y = 30;

my ($l, $r); # FIXME
($l, $r, $txt{w}, $txt{h}) =
	Imager::Font::Wrap->wrap_text(string  => $opts{text},
				      font    => $txt{font},
				      image   => undef,
				      #width  => $img_src->getwidth() - $pad_x,
				      width   => $dst{w} - $pad_x,
			              justify => $opts{justify}) or die Imager->errstr;

# TODO needs to take $opts into account 
($txt{w}, $txt{h}) = ($dst{w}, 
		      $txt{h} + $pad_y);

# TODO cleanup
my $cx = .5 * $pad_x;
my $cy = .5 * $pad_y;

# Create canvas with alpha channel for transparency
$txt{data} = Imager->new(xsize    => $txt{w},
			 ysize    => $txt{h}, 
			 channels => 4) or die Imager->errstr;

# Annotate text
Imager::Font::Wrap->wrap_text(string  => $opts{text},
			      font    => $txt{font},
			      image   => $txt{data},
                              #width  => $img_src->getwidth() - $pad_x,
			      width   => $dst{w} - $pad_x,
			      justify => $opts{justify},
			      x       => $cx,
			      y       => $cy,
			      aa      => 1) or die Imager->errstr;

# Blur
$txt{data}->filter(type      => 'gaussian',
		   stddev    => $opts{text_bg_blur}) or die $txt{data}->errstr;

$txt{data}->filter(type      => 'contrast',
		   intensity => $opts{text_bg_contrast}) or die $txt{data}->errstr;

# Re-annotate 
undef $txt{color};
# Change font color to foreground
$txt{color} = Imager::Color->new(@text_color_fg);
$txt{font}->{color} = $txt{color};
Imager::Font::Wrap->wrap_text(string  => $opts{text},
			      font    => $txt{font},
			      image   => $txt{data},
                              #width  => $img_src->getwidth() - $pad_x,
			      width   => $dst{w} - $pad_x,
			      justify => $opts{justify},
			      x       => $cx,
			      y       => $cy,
			      aa      => 1) or die Imager->errstr;

# Paste (blending alpha channel via rubthrough) text on to main canvas.
# It would be nice if some alignment options were available here:
# top/bottom, for instance.
# TODO: take $opts into account
$offs_x = (.5 * $dst{w}) - (.5 * $txt{w}); # Center
#$offs_y = 0; # Top
$offs_y = $dst{h} - $txt{h}; # Bottom

$dst{data}->rubthrough(src => $txt{data},
		       tx  => $offs_x,
		       ty  => $offs_y) or die Imager->errstr;


$dst{data}->write(file => $dst{file}) or die $dst{data}->errstr;


APOD.pm
#!/usr/bin/env perl
# by gammy
use warnings;
use strict;

=head1 NAME

 APOD - Astronomy Picture Of the Day module.

=head1 DESCRIPTION

 A simple helper to fetch the image and description of NASAs 
 Astronomy Picture Of The Day.
 
=head1 SYNOPSIS

 use APOD;
 my $apod = new APOD;
  
 # Peek at the APOD website to fill in the image url, description, etc.
 # Always call this method first.
 $apod->peek();

 $apod->destination("/tmp/");

 if(-e $apod->destination . '/' . $apod->filename) {
	 die "We already have the APOD!\n";
 }
 
 $apod->save_image();
 $apod->save_description();
 
=head1 Getters

 $url_to_image      = $apod->url;
 $image_data        = $apod->image; # requires ->get_image() or ->save_image()
 $original_filename = $apod->filename;
 $description       = $apod->description;
 $destination_dir   = $apod->description;
 
=head1 Setters

 $new_dest = $apod->destination("/new/destination/");
 
=cut

package APOD;

use base qw/HTML::Parser/;

use constant BASE_URL     => 'http://apod.nasa.gov';

use Carp;
use LWP::Simple;
use HTML::Parser;

my $text_buf;
my $path_found = 0;

# Getters
sub url         { return $_[0]->{url}; }
sub image       { return $_[0]->{image}; }
sub filename    { return $_[0]->{filename}; }
sub description { return $_[0]->{description}; }

# Getter/setter
sub destination { 
	my $self = shift;
	if(@_) {
		$self->{destination} = shift;
	} else {
		if(! $self->{destination}) {
			$self->{destination} = '.';
		}
	}

	return $self->{destination};
};

# Get the APOD webpage and find image url & description. 
# Always call this first.
sub peek {
	my $self = shift;
	my $html = get(BASE_URL) or croak "Failed to get \"" . BASE_URL . "\"";
	$self->parse($html);
}

sub get_image {
	my $self = shift;
	my $url = $self->{url};
	$self->{image} = get($url) or croak "Can't get \"$url\"!";
}

sub save_image {
	my $self = shift;
	my $dst = $self->destination . '/' . $self->{filename};

	croak "Call peek() first!" if ! $self->{description};

	$self->get_image();
	
	# user-defined argument overrides the /entire/ path including 'destination'
	if(@_) { 
		$dst = shift;
	}

	open F, '>', $dst or croak "Can't open \"$dst\": $!";
	print F $self->{image};
	close F;
}

sub save_description {
	my $self = shift;
	my $dst = $self->destination . '/' . $self->{filename} . '.txt';

	croak "Call peek() first!" if ! $self->{description};
	
	# user-defined argument overrides the /entire/ path  including 'destination'
	if(@_) { 
		$dst = shift;
	}

	open F, '>', $dst or croak "Can't open \"$dst\": $!";
	print F $self->{description};
	close F;
}

# Overrides empty HTML::Parser sub
sub start { 
	my ($self, $tagname, $attr, $attrseq, $origtext) = @_;

	if($tagname eq 'p') {
		if($text_buf =~m/Explanation:/s) {
			# The order of these is not random.
			$text_buf =~s/\n+/\n/g;
			$text_buf =~s/\n/ /g;
			$text_buf =~s/ +/ /g;
			$text_buf =~s/^ //g;
			$text_buf =~s/ $//g;
			$text_buf =~s/^Explanation: //;
			$self->{description} = $text_buf;
		
		}
		$text_buf = '';
	}elsif($tagname eq 'a') {
		if($attr->{href}=~m#(image/.*)#s) {
			$self->{url}        = BASE_URL . '/' . $1 if ! $path_found;
			$self->{filename}   = substr($self->{url}, 
			                             rindex($self->{url}, '/') + 1);
			$path_found = 1;
		}
	}
}

# Overrides empty HTML::Parser sub
sub text {
	my ($self, $text) = @_;
	$text_buf .= $text;
}

1;


apod_get.pl
#!/usr/bin/env perl
# 
# - Grab APOD feed (usually contains a weeks worth of items)
# - For each item in the feed,
#   - Grab the APOD page and parse it for the full-size image URL
#   - If we haven't already done so, download the image.
#
# Run this weekly (or twice a week.. or hey why not once per day :p) 
#
# by gammy.

use warnings;
use strict;

use LWP::Simple;
use XML::RSS::Parser;

use constant BASE_URL  => 'http://apod.nasa.gov';
use constant FEED_URL  => BASE_URL . "/apod.rss";
use constant DST_PATH  => $ENV{HOME} . '/.APOD';

if(! -e DST_PATH) {
	mkdir DST_PATH, 0744 or die $!;
}

print "Storing images in \"" . DST_PATH . "\".\n";

my $xml = get(FEED_URL) or die 'Failed to get "' . FEED_URL . '"';

my $parser = new XML::RSS::Parser;
my $feed = $parser->parse_string($xml);

print $feed->item_count . " items.\n";

$| = 1;
for my $item ($feed->query('//item')) {

	my $title = $item->query('title')->text_content;
	my $link = $item->query('link')->text_content;

	print "Get \"$title\"..";
	
	my $tmp = get($link) or die "failed to get \"$link\"";

	# We won't use the 'img src' since this is usually cropped.
	# In case they show a video or something, I assume they don't
	# have them in the "image/"-directory.
	unless($tmp=~m#a href="(image/.+?)"#si) {
		print "contains no image.\n";
		next;
	}

	my $img_url = BASE_URL . "/$1";
	my $filename = substr($1, rindex($1, '/') + 1);

	my $full_path = DST_PATH . "/$filename";

	if(-e $full_path) {
		print "already got it.\n";
		next;
	}

	my $img_data = get($img_url) or die "failed to get \"$img_url\"";

	open F, '>', DST_PATH . "/$filename" or die $!;
	print F $img_data;
	close F;
	
	print "ok.\n";
}

$| = 0;


apod_set.pl
#/usr/bin/env perl
# - Crop image to desktop
# - Annotate message
#
# Programmatical:
# - create empty canvas of desktop size
# - load src image to new canvas
#   - crop image to height of empty canvas if it's taller than desktop
# - paste image to horizontal center of empty canvas
# - create transparent canvas with the annotation text wrapped
#   - apply heavy gaussian blur to "smear" the text out for
#     use as a 'shadow' for the text
#   - increase contrast (does it saturate?)
#   - re-annotate text
# - paste annotation canvas to top horizontal center of canvas
# - store result
#
# Note: The variables need some serious cleanups. It's ridiculous.
# by gammay

use warnings;
use strict;

use Data::Dumper;
use X11::Resolution;
use Imager;
use Imager::Fill;
use Imager::Font::Wrap;
#use Image::Magick;

use constant { 
	SCREEN_NUM  => 0,
	FONT_NAME   => '',
	FONT_PTS    => 14,
	FONT_FILE   => './font/Istok-Regular.ttf',
	BLUR_DEV    => 4.0, # Higher = more smeared background blur
	INTENSITY   => 9.0  # Higher = more contrast on background blur
};

#my @text_bg_color = (64, 64, 128, 255);
#my @text_fg_color = (255, 255, 255, 255);
my @text_bg_color = (10, 10, 16, 255);
my @text_fg_color = (255, 255, 255, 255);

if(@ARGV != 2) {
	die "Usage: $0 <filename> <text>\n";
}

my ($filename, $text) = @ARGV;

if(! -e $filename) {
	die "Can't find \"$filename\"";
}

## Get resolution
my $X = new X11::Resolution;
if($X->{error}){
	die "Can't instantiate X11::Resolution: " . $X->{errorString};
}

my ($w, $h) = $X->getResolution(SCREEN_NUM);

if($X->{error}){
	die "Can't get X resolution: " . $X->{errorString};
}

printf("Resolution of screen %d: %dx%d\n", SCREEN_NUM, $w, $h);

# Create empty canvas
my $canvas_base = Imager->new(xsize    => $w,
			      ysize    => $h,
			      channels => 4) or die Imager->errstr;

# Load source image
my $src = Imager->new(file     => $filename,
		      channels => 4) or die Imager->errstr;

# Scale it to fit height if necessary
if($src->getheight() > $h) {
	my $tmp = $src->scale(ypixels => $h) or die Imager->errstr;
	$src = $tmp;
}

# Paste scaled image to center of canvas
my $offs_x = (.5 * $w) - (.5 * $src->getwidth());
my $offs_y = (.5 * $h) - (.5 * $src->getheight());
$canvas_base->paste(left => $offs_x,
		    top  => $offs_y,
		    src  => $src) or die $canvas_base->errstr;

my $font_color = Imager::Color->new(@text_bg_color);
my $font = Imager::Font->new(file  => FONT_FILE,
			     size  => FONT_PTS,
			     color => $font_color) or die Imager->errstr;

# Calculate size of annotation
# Note that we cannot know the actual width of the text;
# we can only pass our desired maximum width and this is what we get.

# Padding for blurring to dissipate without clipping.
my $pad_x = 30;
my $pad_y = 30;

my ($left, $top, $right, $bottom) =
	Imager::Font::Wrap->wrap_text(string  => $text,
				      font    => $font,
				      image   => undef,
				      width   => ($src->getwidth() - $pad_x),
			              justify => 'left') or die Imager->errstr;

my ($fw, $fh) = ($src->getwidth(), 
		 $bottom + $pad_y);
my $cx = .5 * $pad_x;
my $cy = .5 * $pad_y;

# Create canvas with alpha channel for transparency
my $canvas_text = Imager->new(xsize    => $fw,
			      ysize    => $fh, 
			      channels => 4) or die Imager->errstr;

# Annotate text
Imager::Font::Wrap->wrap_text(string => $text,
			      font   => $font,
			      image  => $canvas_text,
			      width  => $src->getwidth() - $pad_x,
			      x      => $cx,
			      y      => $cy,
			      aa     => 1) or die Imager->errstr;

# Blur
$canvas_text->filter(type => 'gaussian',
		     stddev => BLUR_DEV) or die $canvas_text->errstr;
$canvas_text->filter(type => 'contrast',
		     intensity => INTENSITY) or die $canvas_text->errstr;

# Re-annotate 
undef $font_color;
$font_color = Imager::Color->new(@text_fg_color);
$font->{color} = $font_color;
Imager::Font::Wrap->wrap_text(string => $text,
			      font   => $font,
			      image  => $canvas_text,
			      width  => $src->getwidth() - $pad_x,
			      x      => $cx,
			      y      => $cy,
			      aa     => 1) or die Imager->errstr;

# Paste (blending alpha channel via rubthrough) text on to main canvas.
# It would be nice if some alignment options were available here:
# top/bottom, for instance.
$offs_x = (.5 * $w) - (.5 * $fw); # Center
# $offs_y = 0; # Top
$offs_y = $h - $fh; # Bottom
$canvas_base->rubthrough(src => $canvas_text,
			 tx => $offs_x,
			 ty => $offs_y) or die Imager->errstr;


# Write output
$canvas_base->write(file => "out.png") or die $canvas_base->errstr;


apod_get_today.pl
#!/usr/bin/env perl
# Get full-size APOD as well as description.
# Should be run at least once per day.
#
# It skips the image download if the file already
# exists locally.
#
# by gammy.

use warnings;
use strict;

use constant DST_PATH  => $ENV{HOME} . '/.APOD';

use APOD;

if(! -e DST_PATH) {
	mkdir DST_PATH, 0744 or die $!;
}

my $apod = new APOD;

$apod->peek() or die "$!";

# There should always be a description
die "It doesn't seem like we could parse the page" if ! $apod->description;

# No image? - just die.
die "No image found (perhaps it's a video today).\n" if ! $apod->url;

# Set destination path
$apod->destination(DST_PATH);

my $img_path = $apod->destination . '/' . $apod->filename;
my $dsc_path = "$img_path.txt";

die "Already got \"$img_path\".\n" if -e $img_path;

$apod->save_image();
print "Saved \"$img_path\"\n";

$apod->save_description();
print "Saved \"$dsc_path\"\n";


linnea.pl
#/usr/bin/env perl
# Linnea, a close-minded image annotator.
#
# Other ideas:
# - Clip text to canvas -or- to original image (--clip=desktop, --clip=source)
#
use warnings;
use strict;

use Carp;
use Getopt::Long;

use X11::Resolution;

use Imager;
use Imager::Fill;
use Imager::Font::Wrap;
use Data::Dumper;

my %opts = (
	"text_bg"          => '0a0a10',
	"text_fg"          => 'ffffff',
	"text"             => '',
	"text_file"        => '',
	"font_pts"         => 14,
	"font_file"        => './font/Istok-Regular.ttf',
	"screen"           => 0,
	"size"             => '0x0',
	"desktop-size"     => 0,
	"text_bg_blur"     => 4.0,
	"text_bg_contrast" => 9.0,
	"bg"               => '',
	"valign"           => 'top',    # top center bottom
	"halign"           => 'center', # left center right # TODO
	"justify"          => 'fill',   # left center right fill
);

sub help {

	printf("Usage: %s [args] <input-file> <output-file>
  --text-bg           text backdrop coloring     (default '%s')
  --text-fg           text forgeground color     (default '%s')
  --text              text                       (default '%s')
  --text-file         file containing text       (default '%s')
  --font-pts          point size of font         (default '%s')
  --font-file         font file                  (default '%s')
  --screen            screen number              (default %d)
  --size              output size       (default original size)
  --desktop-size      use desktop size as output size
  --text-bg-blur      bg gaussian blur intensity (default %f)
  --text-bg-contrast  bg contrast intensity      (default %f)
  --bg                bg solid color             (default none)
  --valign            vertical text alignment    (default '%s')
  --halign            horizontal text alignment  (default '%s')
  --justify           text justification method  (default '%s')
", 
	$0,
	$opts{text_bg},
	$opts{text_fg},
	$opts{text},
	$opts{text_file},
	$opts{font_pts},
	$opts{font_file},
	$opts{screen},
	$opts{text_bg_blur},
	$opts{text_bg_contrast},
	$opts{valign},
	$opts{halign},
	$opts{justify});

}

# Input : 6-character hex string (0a1fff) 
# Return: decimal rgb-triplet array 
sub color_parse {
	my $color = lc shift;
	if(length($color) != 6) {
		croak "Bad color '$color': " . 
		      "format is rrggbb where each 'xx'-pair is a hex-value";
	}
	my @colors;
	push @colors, hex substr($color, $_, 2) for (0, 2, 4);

	return @colors;
}

sub resolution_parse {
	my $in = lc shift;
	my @out = split /x/, $in;
	for (@out) {
		croak "Bad size '$in': format is WxH" if /[^\d]/;
	}
	return @out;
}

sub resolution_get {
	my $screen = shift;

	my $X = new X11::Resolution;

	if($X->{error}){
		croak "Can't instantiate X11::Resolution: " . $X->{errorString};
	}

	my ($w, $h) = $X->getResolution($screen);

	if($X->{error}){
		croak "Can't get X resolution: " . $X->{errorString};
	}

	return($w, $h);
}
##############################################################################

my $result = GetOptions ("text-bg=s"          => \$opts{text_bg},
	                 "text-fg=s"          => \$opts{text_fg},
	                 "text=s"             => \$opts{text},
	                 "text-file=s"        => \$opts{text_file},
	                 "font-pts=i"         => \$opts{text_pts},
	                 "font-file=s"        => \$opts{font_file},
	                 "screen=i"           => \$opts{screen},
	                 "size=s"             => \$opts{size},
	                 "desktop-size"       => \$opts{desktop_size},
	                 "text-bg-blur=f"     => \$opts{text_bg_blur},
	                 "text-bg-contrast=f" => \$opts{text_bg_contrast},
	                 "bg=s"               => \$opts{bg},
	                 "valign=s"           => \$opts{valign},
	                 "halign=s"           => \$opts{halign},
	                 "justify=s"          => \$opts{justify},
			 );

if(@ARGV != 2) {
	help();
	exit;
}

# Container hashes
my %src = (
	w    => 0,
	h    => 0,
	data => undef,
);
my %txt = (
	w    => 0,
	h    => 0,
	data => undef,
	color=> undef,
	font => undef,
);
my %dst = (
	file => '',
	w    => 0,
	h    => 0,
	data => undef,
);

($src{file}, $dst{file}) = @ARGV;
my @text_color_fg        = color_parse($opts{text_fg});
my @text_color_bg        = color_parse($opts{text_bg});
my @color_bg;
if($opts{bg} ne '') {
	@color_bg     = color_parse($opts{bg});
}
($dst{w}, $dst{h})       = resolution_parse($opts{size});

if($opts{text_file} ne '') {
	
	if($opts{text} ne '') {
		print "Overriding supplied --text with data in " .
		      "\"$opts{text_file}\"\n";
	}

	$opts{text} = '';

	open F, '<', $opts{text_file} 
		or croak "Can't open \"$opts{text_file}\": $!";

	$opts{text} .= $_ for <F>; 

	close F;

}

if($opts{desktop_size}) {
	($dst{w}, $dst{h}) = resolution_get($opts{screen});
}

##############################################################################

# Load source image
$src{image} = Imager->new(file     => $src{file},
			 channels => 4) or die Imager->errstr;

($src{w}, $src{h}) = ($src{image}->getwidth(), 
		      $src{image}->getheight());

# Set size of dst to src 
if(! $opts{desktop_size}) {
	($dst{w}, $dst{h}) = ($src{w}, $src{h});
}

# Create empty destination canvas
$dst{image} = Imager->new(xsize    => $dst{w},
			 ysize    => $dst{h},
			 channels => 4) or die Imager->errstr;

# Scale it to fit height if necessary
if($src{h} > $dst{h}) {
	my $tmp = $src{image}->scale(ypixels => $dst{h}) or die Imager->errstr;
	$src{image} = $tmp;

	($src{w}, $src{h}) = ($src{image}->getwidth(), $src{image}->getheight());
}

# Set background color if there's a visible canvas
if($src{w} < $dst{w} || $src{h} < $dst{h}) {
	if($opts{bg} ne '') {
		my $color = Imager::Color->new(@color_bg) or die;
		$dst{image}->box(color  => $color,
		                 filled => 1) or die Imager::errstr;

	}
}


# Paste scaled image to center of canvas
# TODO take $opts{halign} into account
my $offs_x = (.5 * $dst{w}) - (.5 * $src{w});
my $offs_y = (.5 * $dst{h}) - (.5 * $src{h});

$dst{image}->paste(left => $offs_x,
		  top  => $offs_y,
		  src  => $src{image}) or die $dst{image}->errstr;

$txt{color} = Imager::Color->new(@text_color_bg);
$txt{font} = Imager::Font->new(file  => $opts{font_file},
			       size  => $opts{font_pts},
			       color => $txt{color}) or die Imager->errstr;

# Calculate size of annotation
# Note that we cannot know the actual width of the text;
# we can only pass our desired maximum width and this is what we get.

# Padding for blurring to dissipate without clipping.
# TODO cleanup
my $pad_x = 30;
my $pad_y = 30;

my ($l, $r); # FIXME
($l, $r, $txt{w}, $txt{h}) =
	Imager::Font::Wrap->wrap_text(string  => $opts{text},
				      font    => $txt{font},
				      image   => undef,
				      #width  => $img_src->getwidth() - $pad_x,
				      width   => $dst{w} - $pad_x,
			              justify => $opts{justify}) or die Imager->errstr;

($txt{w}, $txt{h}) = ($dst{w}, 
		      $txt{h} + $pad_y);

my $cx = .5 * $pad_x;
my $cy = .5 * $pad_y;

# Create canvas with alpha channel for transparency
$txt{image} = Imager->new(xsize    => $txt{w},
			 ysize    => $txt{h}, 
			 channels => 4) or die Imager->errstr;

# Annotate text
Imager::Font::Wrap->wrap_text(string  => $opts{text},
			      font    => $txt{font},
			      image   => $txt{image},
                              #width  => $img_src->getwidth() - $pad_x,
			      width   => $dst{w} - $pad_x,
			      justify => $opts{justify},
			      x       => $cx,
			      y       => $cy,
			      aa      => 1) or die Imager->errstr;

# Blur
$txt{image}->filter(type      => 'gaussian',
		   stddev    => $opts{text_bg_blur}) or die $txt{image}->errstr;

$txt{image}->filter(type      => 'contrast',
		   intensity => $opts{text_bg_contrast}) or die $txt{image}->errstr;

# Re-annotate 
undef $txt{color};
# Change font color to foreground
$txt{color} = Imager::Color->new(@text_color_fg);
$txt{font}->{color} = $txt{color};
Imager::Font::Wrap->wrap_text(string  => $opts{text},
			      font    => $txt{font},
			      image   => $txt{image},
                              #width  => $img_src->getwidth() - $pad_x,
			      width   => $dst{w} - $pad_x,
			      justify => $opts{justify},
			      x       => $cx,
			      y       => $cy,
			      aa      => 1) or die Imager->errstr;

# Paste (blending alpha channel via rubthrough) text on to main canvas.
$offs_x = (.5 * $dst{w}) - (.5 * $txt{w}); # Center

if(lc $opts{valign} eq 'top') {
	$offs_y = 0;
}elsif(lc $opts{valign} eq 'center'){
	$offs_y = (.5 * $dst{h}) - (.5 * $txt{h});
}elsif(lc $opts{valign} eq 'bottom'){
	$offs_y = $dst{h} - $txt{h};
}else{
	die "Invalid valign format. Options are left, center, right.\n";
}

$dst{image}->rubthrough(src => $txt{image},
		       tx  => $offs_x,
		       ty  => $offs_y) or die Imager->errstr;


$dst{image}->write(file => $dst{file}) or die $dst{image}->errstr;
